### PR TITLE
Replace dynamic dispatch with runtime branch on `rev` keyword in sortperm

### LINKED ISSUE #47949: `sortperm!` allocates due to inference/type stability issues
I don't know exactly why `sortperm!` allocates here, but I know that it shouldn't:
```julia
julia> v = rand(10);

julia> ix = similar(eachindex(v));

julia> @btime sortperm!($ix, $v) setup=(copyto!($ix, eachindex($v))) evals=1;
  939.000 ns (5 allocations: 336 bytes)

julia> @btime sort!($ix) setup=(sortperm!($ix, $v)) evals=1;
  84.000 ns (0 allocations: 0 bytes)
```
This is not new in 1.9 but it is slightly worse.

### PR DESCRIPTION
**Pull request by LilithHafner**
This eliminates unexpected allocations in `sortperm` and `sortperm!`, sometimes substantially improving runtime for small inputs:

```julia
julia> v = rand(10); ix = similar(eachindex(v));

julia> @btime sortperm!(copyto!($ix, eachindex($v)), $v);
  571.896 ns (3 allocations: 48 bytes) => 106.496 ns (0 allocations: 0 bytes)

julia> @btime sortperm($v);
  598.118 ns (4 allocations: 192 bytes) => 139.678 ns (1 allocation: 144 bytes)
```

Fixes #47949.

---

Outdated OP:

~I think this might avoid compiler heuristics that arise due to keyword arguments in sortperm, but I still don't understand why the allocations were there in the first place. Fixes #47949.~

~- add strange hacks to avoid compiler heuristics?~
~- add allocation tests~

~The situation is still not perfect (e.g. `@test_broken 0 == @allocations sortperm!(i, v, rev=true)`), but it is better according to allocation tests and some benchmarks.~

### CODE REVIEW

**Comment by LilithHafner** [comment]
@nanosoldier `runbenchmarks("sort", vs=":master")`

**Comment by LilithHafner** [comment]
This improves the situation with #47811, but a regression remains for sizes 32–256.

```julia
julia> for i in 1:12
           n = 2^i
           print(lpad(n, 4)); @btime sortperm(x) setup=(x=rand($n));
       end
   2  224.951 ns (2 allocations: 96 bytes)  => 481.651 ns (4 allocations: 128 bytes) => 52.206 ns (1 allocation: 80 bytes)
   4  236.767 ns (2 allocations: 112 bytes) => 494.113 ns (4 allocations: 144 bytes) => 64.007 ns (1 allocation: 96 bytes)
   8  261.952 ns (2 allocations: 144 bytes) => 523.632 ns (4 allocations: 176 bytes) => 87.860 ns (1 allocation: 128 bytes)
  16  315.431 ns (2 allocations: 208 bytes) => 726.031 ns (5 allocations: 432 bytes) => 246.363 ns (2 allocations: 384 bytes)
  32  482.479 ns (2 allocations: 352 bytes) => 1.218 μs (5 allocations: 720 bytes)   => 684.515 ns (2 allocations: 672 bytes)
  64  858.757 ns (2 allocations: 592 bytes) => 2.011 μs (5 allocations: 1.17 KiB)    => 1.609 μs (2 allocations: 1.12 KiB)
 128  2.150 μs (2 allocations: 1.08 KiB)    => 3.818 μs (5 allocations: 2.17 KiB)    => 3.446 μs (2 allocations: 2.12 KiB)
 256  5.314 μs (2 allocations: 2.14 KiB)    => 7.808 μs (5 allocations: 4.30 KiB)    => 7.536 μs (2 allocations: 4.25 KiB)
 512  15.903 μs (2 allocations: 4.14 KiB)   => 15.624 μs (5 allocations: 8.30 KiB)   => 15.077 μs (2 allocations: 8.25 KiB)
1024  45.238 μs (2 allocations: 8.14 KiB)   => 32.604 μs (5 allocations: 16.30 KiB)  => 33.118 μs (2 allocations: 16.25 KiB)
2048  99.861 μs (2 allocations: 16.14 KiB)  => 69.427 μs (5 allocations: 32.30 KiB)  => 70.779 μs (2 allocations: 32.25 KiB)
4096  219.132 μs (3 allocations: 32.06 KiB) => 150.960 μs (7 allocations: 64.14 KiB) => 152.937 μs (4 allocations: 64.09 KiB)

julia> VERSION
v"1.8.3" => v"1.10.0-DEV.181" => PR
```

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47966#issuecomment-1362773387) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/bd36d0a_vs_12e679c/report.md).

**Comment by LilithHafner** [comment]
Performance diff vs master on https://github.com/JuliaCI/BaseBenchmarks.jl/pull/305 is free of regressions and shows some improvements
```
"insertionsort", "sortperm! reverse" => TrialJudgement(-23.35% => improvement)
"insertionsort", "sortperm forwards" => TrialJudgement(-18.31% => invariant)
"length = 3", "sortperm(rand(length))" => TrialJudgement(-80.03% => improvement)
"length = 10", "sortperm(rand(length))" => TrialJudgement(-68.33% => improvement)
"length = 10", "mixed eltype with by order" => TrialJudgement(-20.71% => improvement)
"length = 30", "sortperm(rand(length))" => TrialJudgement(-44.86% => improvement)
"length = 100", "sortperm(rand(length))" => TrialJudgement(-13.21% => invariant)
"issues", "sortperm on a view (Float64)" => TrialJudgement(-9.12% => invariant)
```

**Comment by LilithHafner** [comment]
> [Your benchmark job](https://github.com/JuliaLang/julia/pull/47966#issuecomment-1362773387) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/bd36d0a_vs_12e679c/report.md).

As expected, `sortperm!` allocations go to 0 for insertion & quick sorts

@nanosoldier `runbenchmarks("sort", vs = ":release-1.8")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47966#issuecomment-1362830215) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/bd36d0a_vs_61c0b41/report.md).

**Comment by LilithHafner** [comment]
@nanosoldier `runbenchmarks("sort", vs = ":release-1.8")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47966#issuecomment-1363316743) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4d258ca_vs_61c0b41/report.md).

**Comment by LilithHafner** [comment]
There are no remaining regressions vs 1.8 in BaseBenchmarks' `sortperm` and `sorpterm!` benchmarks!

Some `sort` regressions are still detected and some `sortperm` regressions remain and are detected by the more diverse https://github.com/JuliaCI/BaseBenchmarks.jl/pull/305 benchmarks.

**Comment by LilithHafner** [comment]
Bump

**Comment by oscardssmith** [comment]
ready to merge?

**Comment by LilithHafner** [comment]
I think so; I'm not entirely sure _how_ this PR improves the situation, but I'm pretty confident that it does. I imagine compiler heuristics about inlining and constant propagation are involved.

I'd be comfortable with merging this as is, but it would be even better to have someone who understands the exact mechanisms at play here.

File: base/sort.jl [hunk 1]
```diff
@@ -1625,14 +1631,26 @@ function sortperm!(ix::AbstractArray{T}, A::AbstractArray;
...
+        sort!(ix; alg, order=Perm(ord(lt, by, true, order), vec(A)), scratch, dims...)
+    else
+        sort!(ix; alg, order=Perm(ord(lt, by, nothing, order), vec(A)), scratch, dims...)
     end
-    sort!(ix; alg, order = Perm(ord(lt, by, rev, order), vec(A)), scratch, dims...)
 end
 
+# TODO stop using these three hacks
+# but check performance, especially unexpected allocations, when removing
+Base.@assume_effects :nothrow very_unsafe_copyto!(a, b) = a .= b
```
Comment (oscardssmith) [inline base/sort.jl:1649]: this is scary

Comment (LilithHafner) [reply base/sort.jl:1649]: Yeah... @aviatesk, is this sort of thing acceptable or is there another way?

Comment (N5N3) [reply base/sort.jl:1649]: My local test shows that perhaps `@noinline` is enough?

Comment (LilithHafner) [reply base/sort.jl]: This is why we do code review. Thanks!

I think I introduced these when they were necessary and then added branching on `rev` which is more impactful and fixes the root of the issue so now none of these hacks are necessary at all, and CI checks for unexpected allocations automatically :)

**Comment by LilithHafner** [comment]
@nanosoldier `runbenchmarks("sort", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47966#issuecomment-1366090346) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/f2b4ff5_vs_a2db90f/report.md).

**Comment by KristofferC** [comment]
This is good to go now, right?

**Comment by LilithHafner** [comment]
Comparing to before the recent changes, when performance was good to go but there were still unnecessary unsafe operations.

@nanosoldier `runbenchmarks("sort", vs="@4191e27f71ee7a32e5e6b2a932603ce341070ed1")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47966#issuecomment-1366188304) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/07b703b_vs_4191e27/report.md).

**Comment by LilithHafner** [comment]
lgtm

**Pull request merged**